----
# **<span style="color:darkmagenta">PROYECTO EDA - BANK MARKETING</span>**
----

---
---
## <span style="color:gray">**1. Importación de librerías**</span> 📂

In [ ]:
# Tratamiento de datos
import numpy as np
import pandas as pd 

# Visualización de datos
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap

---
---
## <span style="color:gray">**2. Carga de datos**</span> 📥

In [201]:
# Datos CSV
bank_data = pd.read_csv("../data/raw/bank-additional.csv")

# Datos Excel
customer_data = pd.read_excel("../data/raw/customer-details.xlsx", sheet_name=None, index_col=0) # Devuelve un diccionario con todas las hojas

In [202]:
# Unión de las hojas de excel en un solo DataFrame
customer_data = pd.concat(customer_data.values(), ignore_index=True)

In [203]:
# Configuración para mostrar todas las columnas
pd.set_option('display.max_columns', None)

---
---


----
# <span style="color:darkmagenta">**Desarrollo del proyecto**</span>
----

---
---
## <span style="color:gray">**Exploración inicial de los datasets**</span> 🔎

Antes de comenzar el análisis exploratorio de datos (EDA), es importante realizar una inspección general de los dataset. Este primer contacto con los datos nos permite entender su estructura, el tipo de variables que contienen y detectar posibles problemas como valores faltantes o formatos incorrectos.

Para ello utilizamos varias funciones de Pandas que nos ayudan a obtener una visión rápida y completa del conjunto de datos.

---

## `bank_data`

👀 **Visualización de las primeras y últimas filas**

In [204]:
bank_data.head(3)

,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
0,0,NaN,housemaid,MARRIED,basic.4y,0.0,0.0,0.0,telephone,261,1,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,2-agosto-2019,41.495,-71.233,089b39d8-e4d0-461b-87d4-814d71e0e079
1,1,57.0,services,MARRIED,high.school,NaN,0.0,0.0,telephone,149,1,999,0,NONEXISTENT,1.1,"93,994","-36,4",NaN,5191,no,14-septiembre-2016,34.601,-83.923,e9d37224-cb6f-4942-98d7-46672963d097
2,2,37.0,services,MARRIED,high.school,0.0,1.0,0.0,telephone,226,1,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,15-febrero-2019,34.939,-94.847,3f9f49b5-e410-4948-bf6e-f9244f04918b


In [205]:
bank_data.tail(3)

,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
42997,15046,NaN,blue-collar,SINGLE,basic.6y,0.0,1.0,0.0,cellular,391,2,999,0,NONEXISTENT,1.4,"93,918","-42,7",NaN,"5228,1",no,15-septiembre-2016,40.679,-120.015,cadadd4b-7ee5-4019-b13a-ca01bb67ca5b
42998,15280,NaN,admin.,MARRIED,university.degree,NaN,0.0,0.0,cellular,674,3,999,0,NONEXISTENT,1.4,"93,918","-42,7","4,958","5228,1",no,23-septiembre-2019,27.772,-117.518,5f432048-d515-4bb5-9c94-62db451f88d4
42999,27570,NaN,unemployed,SINGLE,university.degree,0.0,0.0,1.0,cellular,104,2,999,0,NONEXISTENT,-0.1,"93,2",-42,"4,021","5195,8",no,6-noviembre-2019,41.146,-105.026,993bbbd6-4dbc-4a40-a408-f91f8462bee6


🎲 **Visualización de filas aleatorias**

In [206]:
bank_data.sample(3)

,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
33740,33884,31.0,admin.,MARRIED,high.school,0.0,1.0,0.0,cellular,476,1,999,0,NONEXISTENT,-1.8,"92,893","-46,2","1,281","5099,1",yes,27-febrero-2017,24.686,-93.705,24911894-2287-437e-a226-913334dfbbf8
27375,27519,48.0,unemployed,MARRIED,professional.course,NaN,1.0,1.0,telephone,83,2,999,0,NONEXISTENT,-0.1,"93,2",-42,"4,021","5195,8",no,22-marzo-2017,39.140,-86.137,20ea5caf-06c1-4c08-bb71-7c0aa0804365
38923,39090,35.0,blue-collar,MARRIED,basic.9y,0.0,0.0,1.0,cellular,520,3,999,2,FAILURE,-3.0,"92,713",-33,NaN,"5023,5",no,20-noviembre-2017,32.072,-78.443,c6701bc5-b7f8-4f0b-9ec6-2eccaf15f287


*La inspección inicial muestra posibles inconsistencias en el formato de los datos, incluyendo diferencias en el uso de mayúsculas/minúsculas y signos de puntuación. Esto sugiere que será necesario aplicar procesos de limpieza y estandarización en etapas posteriores del análisis. Además, la columna `Unnamed: 0` parece corresponder a un índice generado automáticamente al exportar el dataset, por lo que habría que investigar si conviene eliminarla. Y la columna `previous` parece que siempre es 0, lo cuál habría que investigar.* 

📏 **Dimensión del dataset**

In [207]:
bank_data.shape

(43000, 24)

*Este DataFrame está compuesto do 43000 filas y 24 columnas.*

🧾 **Información general**

In [208]:
bank_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  str    
 3   marital         42915 non-null  str    
 4   education       41193 non-null  str    
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  str    
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  str    
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  str    
 16  cons.conf.idx   43000 non-null  str    
 17  euribor3m       33744 non-null  str    
 1

*A partir de la información proporcionada por info(), se observa que algunas columnas presentan valores nulos. Además, ciertos campos parecen tener un tipo de dato incorrecto, lo que sugiere la necesidad de realizar procesos de limpieza y conversión de tipos antes de continuar con el análisis.*

❌ **Detección de nulos**

In [209]:
nulos_bank = bank_data.isnull().sum()
print(f"El número de valores nulos de cada columna es: \n{nulos_bank[nulos_bank > 0]}")

El número de valores nulos de cada columna es: 
age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
cons.price.idx     471
euribor3m         9256
date               248
dtype: int64


In [210]:
nulos_bank_porcentaje = bank_data.isna().sum() / (bank_data.shape[0])*100
print(f"El porcentaje de valores nulos de cada columna es: \n{nulos_bank_porcentaje[nulos_bank_porcentaje > 0].round(2)}")

El porcentaje de valores nulos de cada columna es: 
age               11.91
job                0.80
marital            0.20
education          4.20
default           20.89
housing            2.39
loan               2.39
cons.price.idx     1.10
euribor3m         21.53
date               0.58
dtype: float64


*Algunas columnas presentan un alto porcentaje de valores nulos; será necesario evaluar posteriormente cómo manejarlos (eliminación, imputación u otro tratamiento) antes de continuar con el análisis.*

🔁 **Detección de duplicados**

In [211]:
bank_data.duplicated().sum()

np.int64(0)

*No hay valores duplicados*

🏷️ **Nombres de las columnas**

In [212]:
bank_data.columns

Index(['Unnamed: 0', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'latitude',
       'longitude', 'id_'],
      dtype='str')

*La revisión de los nombres de las columnas muestra que algunas variables no describen claramente la información que contienen o incluyen caracteres innecesarios.
Para mejorar la legibilidad y consistencia del dataset, se realizará un proceso de estandarización y renombrado de columnas.*

📊 **Estadísticas descriptivas**

In [213]:
bank_data.describe().round(2).T

,count,mean,std,min,25%,50%,75%,max
Unnamed: 0,43000.0,20651.10,11868.77,0.0,10381.75,20642.50,30930.25,41187.00
age,37880.0,39.98,10.44,17.0,32.00,38.00,47.00,98.00
default,34019.0,0.00,0.01,0.0,0.00,0.00,0.00,1.00
housing,41974.0,0.54,0.50,0.0,0.00,1.00,1.00,1.00
loan,41974.0,0.16,0.36,0.0,0.00,0.00,0.00,1.00
duration,43000.0,257.74,258.67,0.0,102.00,179.00,319.00,4918.00
campaign,43000.0,2.57,2.77,1.0,1.00,2.00,3.00,56.00
pdays,43000.0,962.33,187.26,0.0,999.00,999.00,999.00,999.00
previous,43000.0,0.17,0.50,0.0,0.00,0.00,0.00,7.00
emp.var.rate,43000.0,0.08,1.57,-3.4,-1.80,1.10,1.40,1.40


*Este análisis permite entender la distribución, rangos de valores y detectar posibles anomalías en las variables numéricas:*

   - *La variable **age** presenta una media cercana a 40 años, con valores entre 17 y 98 años.*  
   - *La duración de las llamadas (**duration**) muestra una alta dispersión y valores extremos, lo que indica posibles outliers.*
   - *Variables como **campaign** y **previous** muestran que la mayoría de los clientes han sido contactados pocas veces.* 
   - *La variable **pdays** tiene gran concentración de valores en 999, lo que probablemente representa clientes no contactados previamente.*
   - *El valor máximo de la columna `Unnamed: 0` no llega al número total de filas, lo cual podría indicar duplicados o nulos. Si es así, habría que eliminarla antes de continuar con el análisis dado que no aporta información relevante.*

In [214]:
bank_data.describe(include=['category', 'str']).T

,count,unique,top,freq
job,42655,11,admin.,10873
marital,42915,3,MARRIED,25999
education,41193,7,university.degree,12722
contact,43000,2,cellular,27396
poutcome,43000,3,NONEXISTENT,37103
cons.price.idx,42529,26,"93,994",7938
cons.conf.idx,43000,26,"-36,4",8020
euribor3m,33744,309,"4,857",2287
nr.employed,43000,11,"5228,1",16980
y,43000,2,no,38156


*Este análisis permite identificar la distribución de las variables categóricas, así como posibles desbalances y necesidades de transformación:*

- *Las variables **job**, **marital** y **education** presentan múltiples categorías con cierta concentración en valores específicos.*
- *La variable **contact** está dominada por cellular, lo que refleja que la mayoría de los contactos se realizan por teléfono móvil.*
- *La variable **poutcome** presenta una fuerte concentración en NONEXISTENT, lo que indica que la mayoría de los clientes no participaron en campañas previas.*
- *Las variables **cons.price.idx**, **cons.conf.idx**, **euribor3m** y **nr.employed** actualmente están siendo tratadas como categóricas y habría que convertirlas a tipo numérico.*
- *La variable objetivo **y** está claramente desbalanceada, con una mayoría de respuestas no.*
- *La variable **date** contiene un gran número de valores únicos, lo que indica alta variabilidad temporal y la necesidad de transformarla en variables más informativas (mes, año).*

### Conclusiones `bank_data`

A partir de la exploración inicial del dataset se pueden destacar las siguientes observaciones:

   - El dataset contiene aproximadamente 43 000 registros.
   - Se identifican valores faltantes en algunas variables.
   - Algunas columnas presentan inconsistencias en su formato.
   - Las variables categóricas `job`, `marital`, `education`, `contact` y `poutcome`están desbalanceadas.
   - Se observan posibles outliers en `duration` y concentraciones en `pdays`, `campaign`, `previous`.
   - La variable objetivo `y` presenta un claro predominio de no.
   - La columna `date` requiere transformación a variables temporales.
   - La columna `Unnamed: 0` podría contener duplicados o nulos. 

---
## `customer_data`

👀 **Visualización de las primeras y últimas filas**

In [215]:
customer_data.head(3)

,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
0,161770,1,0,2012-04-04,29,089b39d8-e4d0-461b-87d4-814d71e0e079
1,85477,1,1,2012-12-30,7,e9d37224-cb6f-4942-98d7-46672963d097
2,147233,1,1,2012-02-02,5,3f9f49b5-e410-4948-bf6e-f9244f04918b


In [216]:
customer_data.tail(3)

,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
43167,78470,2,2,2014-10-28,12,d6271666-319d-42c8-a741-cb22bf2c2093
43168,144218,0,1,2014-07-01,9,f67fbfa8-6573-414d-a805-b26a2f1b1ceb
43169,115155,0,2,2014-12-25,7,9727bc06-c11a-461a-a5bb-3d210467cc2a


🎲 **Visualización de filas aleatorias**

In [217]:
customer_data.sample(3)

,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
39241,109069,2,1,2014-05-02,10,9d065aab-f92b-41ed-8c6e-c47971cf0e0c
19975,169473,1,1,2012-10-18,7,b97eefea-c5bc-4bc8-8685-4cea97c574dc
9007,121565,0,2,2012-08-30,8,c1f8d747-6a1a-401b-8155-90fbe06733ca


📏 **Dimensión del dataset**

In [218]:
customer_data.shape

(43170, 6)

🧾 **Información general**

In [219]:
customer_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Income             43170 non-null  int64         
 1   Kidhome            43170 non-null  int64         
 2   Teenhome           43170 non-null  int64         
 3   Dt_Customer        43170 non-null  datetime64[us]
 4   NumWebVisitsMonth  43170 non-null  int64         
 5   ID                 43170 non-null  str           
dtypes: datetime64[us](1), int64(4), str(1)
memory usage: 2.0 MB


❌ **Detección de nulos**

In [220]:
customer_data.isnull().sum()

Income               0
Kidhome              0
Teenhome             0
Dt_Customer          0
NumWebVisitsMonth    0
ID                   0
dtype: int64

🔁 **Detección de duplicados**

In [221]:
customer_data.duplicated().sum()

np.int64(0)

📊 **Estadísticas descriptivas**

In [222]:
customer_data.describe().round(2).T

,count,mean,min,25%,50%,75%,max,std
Income,43170.0,93227.39,5841.0,49608.0,93009.5,136740.5,180802.0,50498.18
Kidhome,43170.0,1.0,0.0,0.0,1.0,2.0,2.0,0.82
Teenhome,43170.0,1.0,0.0,0.0,1.0,2.0,2.0,0.82
Dt_Customer,43170,2013-05-11 22:09:47.407922,2012-01-01 00:00:00,2012-07-16 00:00:00,2013-03-01 00:00:00,2014-03-28 00:00:00,2014-12-31 00:00:00,NaN
NumWebVisitsMonth,43170.0,16.59,1.0,9.0,17.0,25.0,32.0,9.24


### Conclusiones `customer_data`

A partir de la exploración inicial del dataset se pueden destacar las siguientes observaciones:

- El dataset contiene 43170 registros y 6 variables, sin presencia de valores nulos ni duplicados.
- La variable `Income` presenta una alta variabilidad, con valores entre 5841 y 180802, lo que sugiere diferencias significativas en el nivel de ingresos y posibles valores extremos.
- Las variables `Kidhome` y `Teenhome` toman valores entre 0 y 2, con medias cercanas a 1, indicando que la mayoría de los clientes tiene uno o ningún hijo en casa.
- La variable `NumWebVisitsMonth` muestra una media de aproximadamente 16 visitas mensuales, con un rango entre 1 y 32, lo que refleja distintos niveles de interacción digital entre los clientes.